libraries/AP_Proximity/AP_Proximity_LightWareSF40C.cpp

```diff
@@ -51,62 +57,87 @@ void AP_Proximity_LightWareSF40C::update(void)
-    if ((_last_distance_received_ms == 0) || (AP_HAL::millis() - _last_distance_received_ms > PROXIMITY_SF40C_TIMEOUT_MS)) {
+    if ((last_distance_received_ms == 0) || ((AP_HAL::millis() - last_distance_received_ms) > PROXIMITY_SF40C_TIMEOUT_MS)) {
         set_status(AP_Proximity::Proximity_NoData);
     } else {
         set_status(AP_Proximity::Proximity_Good);
     }
 }
 
-// get maximum and minimum distances (in meters) of primary sensor
-float AP_Proximity_LightWareSF40C::distance_max() const
-{
-    return 100.0f;
-}
-float AP_Proximity_LightWareSF40C::distance_min() const
-{
-    return 0.20f;
-}
-
 // initialise sensor (returns true if sensor is succesfully initialised)
-bool AP_Proximity_LightWareSF40C::initialise()
+void AP_Proximity_LightWareSF40C::initialise()
 {
-    // set motor direction once per second
-    if (_motor_direction > 1) {
-        if ((_last_request_ms == 0) || AP_HAL::millis() - _last_request_ms > 1000) {
-            set_motor_direction();
-        }
+    // initialise sectors
+    if (!sector_initialised) {
+        init_sectors();
     }
-    // set forward direction once per second
-    if (_forward_direction != frontend.get_yaw_correction(state.instance)) {
-        if ((_last_request_ms == 0) || AP_HAL::millis() - _last_request_ms > 1000) {
-            set_forward_direction();
-        }
+
+    // exit immediately if we've sent initialisation requests in the last second
+    uint32_t now_ms = AP_HAL::millis();
+    if ((AP_HAL::millis() - last_request_ms) < 1000) {
+        return;
     }
-    // request motors turn on once per second
-    if (_motor_speed == 0) {
-        if ((_last_request_ms == 0) || AP_HAL::millis() - _last_request_ms > 1000) {
-            set_motor_speed(true);
-        }
-        return false;
+    last_request_ms = now_ms;
+
+    // if no replies in last 15 seconds reboot sensor
+    if ((now_ms > 30000) && (now_ms - last_reply_ms > 15000)) {
+        restart_sensor();
     }
-    // initialise sectors
-    if (!_sector_initialised) {
-        init_sectors();
-        return false;
+
+    // re-fetch motor state
+    request_motor_state();
+
+    // get token from sensor (required for reseting)
+    if (!got_token()) {
+        request_token();
+        return;
+    }
+
+    // if motor is starting up give more time to succeed or fail
+    if (sensor_state.motor_state < 3) {
+        return;
+    }
+
+    // if motor fails, reset sensor and re-try everything
+    if (sensor_state.motor_state >= 4) {
+        restart_sensor();
+        return;
+    }
+
+    // motor is running correctly (motor_state is 3) so request start of streaming
+    if (!sensor_state.streaming || (sensor_state.output_rate != PROXIMITY_SF40C_DESIRED_OUTPUT_RATE)) {
+        request_stream_start();
+        return;
+    }
+
+    // if we haven't received any distance messages in 15 seconds restart sensor and re-try everything
+    if ((now_ms - last_distance_received_ms) > 15000) {
```

Comment:
I think it's slightly different than the first reboot check because this checks if any **distances** have been received while the first checks if any responses at all have been received.  Some of this was me trying to handle the case where the lidar loses power to it's motor (which is powered from the 12V battery) but still has power to it's CPU (which is powered from the flight controller). In testing though it seems that sending the restart command to the sensor's CPU will not cause it to recover when the 12V power is restored so I'll just remove this. txs